neps/space/neps_spaces/config_string.py

```diff
@@ -0,0 +1,327 @@
+        result.append(value)
+    result.append(")" * current_level)
+
+    result_string = "".join(result).strip()
+    if result_string.startswith("("):
+        result_string = result_string[1:-1].strip()
+
+    # A workaround needed since in the existing configurations
+    #  generated by previous methods, e.g. the `resBlock resBlock` and `resBlock` items
+    #  occur without wrapping parenthesis, differently from other items.
+    # Wrap them appropriately in parentheses here and in the inverse process.
+    # For example 'id' comes in two forms: 'id id' and 'Ops id',
+    #  only the 'id id' variant should be replaced.
+    replacements = [
+        ("resBlock", True),
+        ("id", False),
+    ]
+    for op, replace_individual in replacements:
+        result_string = result_string.replace(f"({op} {op})", "__TMP_PLACEHOLDER___")
+        if replace_individual:
+            result_string = result_string.replace(f"({op})", f"{op}")
+        result_string = result_string.replace("__TMP_PLACEHOLDER___", f"{op} {op}")
+    return result_string
+
+
+class ConfigString:
+    """A class representing a configuration string in NePS spaces.
+    It provides methods to unwrap the configuration string into structured parts,
+    retrieve the maximum hierarchy level, and get a representation of the configuration
+    at a specific hierarchy level.
+    """
+
+    def __init__(self, config_string: str) -> None:
+        """Initialize the ConfigString with a given configuration string.
+
+        Args:
+            config_string: The configuration string to be wrapped.
+
+        Raises:
+            ValueError: If the config_string is None or empty.
+        """
+        if config_string is None or len(config_string) == 0:
+            raise ValueError(f"Invalid config string: {config_string}")
+        self.config_string = config_string
+
+        # The fields below are needed for lazy and cached evaluation.
+        # In python 3.8+ can be replaced by `cached_property`
+        self._unwrapped: tuple[UnwrappedConfigStringPart, ...] | None = None
+        self._max_hierarchy_level: int | None = None
+
+        # a cache for the different hierarchy levels of this config string
+        self._at_hierarchy_level_cache: dict[int, ConfigString] = {}
+
+    @property
+    def unwrapped(self) -> tuple[UnwrappedConfigStringPart, ...]:
+        """Get the unwrapped representation of the configuration string.
+
+        Returns:
+            A tuple of UnwrappedConfigStringPart objects representing the unwrapped
+            config.
+
+        Raises:
+            ValueError: If there is an error unwrapping the config string.
+        """
+        # If the unwrapped is already cached, return it
+        if self._unwrapped is not None:
+            return self._unwrapped
+
+        unwrapped = unwrap_config_string(self.config_string)
+        if not unwrapped:
+            raise ValueError(f"Error unwrapping config string: {self.config_string}")
```

Comment:
This message should be improved to propagate and display the root-cause error that led to the program's failure